keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,494 @@
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Logging Interface](#logging-interface)
+    - [Logger initialization](#logger-initialization)
+    - [Storing Logger and metadata in Context](#storing-logger-and-metadata-in-context)
+    - [Example](#example)
+  - [New logging formats](#new-logging-formats)
+    - [Production format](#production-format)
+    - [Development format](#development-format)
+  - [Logging configuration](#logging-configuration)
+  - [Kubernetes log schema](#kubernetes-log-schema)
+  - [Migration](#migration)
+    - [Stage 1 / Alpha](#stage-1--alpha)
+    - [Preparation](#preparation)
+    - [Automatic migration](#automatic-migration)
+    - [Stage 2 / Beta](#stage-2--beta)
+    - [Stage 3 / GA](#stage-3--ga)
+    - [Stage 4 / Deprecation](#stage-4--deprecation)
+  - [User Stories](#user-stories)
+    - [Story 1 - Kubernetes developer working with local logs](#story-1---kubernetes-developer-working-with-local-logs)
+    - [Story 2 - Cluster Administrator configures log ingestion into ElasticSearch](#story-2---cluster-administrator-configures-log-ingestion-into-elasticsearch)
+    - [Story 3 - Administrator uses Elasticsearch to debug problem with workload](#story-3---administrator-uses-elasticsearch-to-debug-problem-with-workload)
+  - [Performance](#performance)
+  - [Risks and Mitigations](#risks-and-mitigations)
+    - [Migration failing halfway](#migration-failing-halfway)
+    - [Hard to customize logger format](#hard-to-customize-logger-format)
+    - [Huge increase of log volume](#huge-increase-of-log-volume)
+- [Design Details](#design-details)
+  - [Mapping between logging methods](#mapping-between-logging-methods)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Alpha](#alpha)
+    - [Beta](#beta)
+    - [GA](#ga)
+    - [Deprecating old logging format](#deprecating-old-logging-format)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+    - [Enhance klog instead of replacing it with logr](#enhance-klog-instead-of-replacing-it-with-logr)
+    - [Avoid double writing of logs using klog and logr APIs](#avoid-double-writing-of-logs-using-klog-and-logr-apis)
+- [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+- [Production Readiness Review Process](#production-readiness-review-process)
+<!-- /toc -->
+
+## Summary
+
+Current logging in the Kubernetes control-plane doesn’t provide any uniform structure for logs which makes storing, processing, querying and analyzing logs hard and forces administrators and developers to rely on ad-hoc solutions. This KEP proposes a new logging API and a mechanism for standardizing logging keys that makes logs much easier to ingest and process by modern logging solutions. 
```

Comment:
Link to previously reported issue: https://github.com/kubernetes/kubernetes/issues/69825